File: backend/api/app/data/repositories/sketch.py
from typing import Iterable

from app.data.datasources.local.sketch import SketchLocalDataSource
from app.domain.entities.sketch import Sketch, SketchEntity
from app.domain.repositories.sketch import BaseRepository
from core.common.either import Either
from core.errors.exceptions import CacheException
from core.errors.failure import CacheFailure, Failure


class SketchRepositoryImpl(BaseRepository):
    
    def __init__(self, sketch_local_datasource: SketchLocalDataSource):
        self.sketch_local_datasource = sketch_local_datasource

    async def create_sketch(self, sketch: Sketch, team_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        try:
            sketch_entity = await self.sketch_local_datasource.create_sketch(team_id=team_id, user_id=user_id, sketch=sketch)
            return Either.right(sketch_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def update_sketch(self, sketch: Sketch, sketch_id: str, team_id: str,  user_id: str) -> Either[Failure, SketchEntity]:
        try:
            sketch_entity = await self.sketch_local_datasource.update_sketch(team_id=team_id, user_id=user_id, sketch=sketch, sketch_id=sketch_id)
            return Either.right(sketch_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def delete_sketch(self, sketch_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        try:
            sketch_entity = await self.sketch_local_datasource.delete_sketch(sketch_id=sketch_id, user_id=user_id)
            return Either.right(sketch_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def views_sketch(self, team_id: str, user_id: str) -> Either[Failure, Iterable[SketchEntity]]:
        try:
            sketch_entity = await self.sketch_local_datasource.views_sketch(team_id=team_id, user_id=user_id)
            return Either.right(sketch_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

    async def view_sketch(self, sketch_id: str, user_id: str) -> Either[Failure, SketchEntity]:
        try:
            sketch_entity = await self.sketch_local_datasource.view_sketch(sketch_id=sketch_id, user_id=user_id)
            return Either.right(sketch_entity)
        except CacheException as e:
            return Either.left(CacheFailure(error_message=str(e)))

File: mobile/lib/features/architect/data/repositories/sketch.dart
import 'package:architect/core/errors/exception.dart';
import 'package:architect/features/architect/domains/entities/sketch.dart';
import 'package:dartz/dartz.dart';

import '../../../../core/errors/failure.dart';
import '../../../../core/network/network_info.dart';
import '../../domains/repositories/sketch.dart';
import '../datasources/local/auth.dart';
import '../datasources/remote/sketch.dart';

class SketchRepositoryImpl implements SketchRepository {
  SketchRepositoryImpl({
    required this.remoteDataSource,
    required this.authLocalDataSource,
    required this.networkInfo,
  });

  final RemoteSketchDataSource remoteDataSource;
  final AuthLocalDataSource authLocalDataSource;
  final NetworkInfo networkInfo;

  @override
  Future<Either<Failure, Sketch>> create({
    required String title,
    required String teamId,
  }) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final sketch = await remoteDataSource.create(
            title: title, teamId: teamId, token: auth.accessToken);
        return Right(sketch);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Sketch>> delete(String sketchId) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final sketch = await remoteDataSource.delete(
            sketchId: sketchId, token: auth.accessToken);
        return Right(sketch);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Sketch>> update(
      {required String sketchId,
      required String title,
      required String teamId}) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final sketch = await remoteDataSource.update(
            sketchId: sketchId,
            title: title,
            teamId: teamId,
            token: auth.accessToken);

        return Right(sketch);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, Sketch>> view(String sketchId) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final sketch = await remoteDataSource.view(sketchId, auth.accessToken);
        return Right(sketch);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }

  @override
  Future<Either<Failure, List<Sketch>>> views({required String teamId}) async {
    if (await networkInfo.isConnected) {
      try {
        final auth = await authLocalDataSource.getToken();
        final sketches = await remoteDataSource.views(
            teamId: teamId, token: auth.accessToken);
        return Right(sketches);
      } on ServerException {
        return Left(ServerFailure());
      }
    } else {
      return Left(ServerFailure());
    }
  }
}
